### ISSUE TITLE
[explicit-module-boundary-types] Complains about inner function expressions

### ISSUE THREAD

**Issue by billyjanitsch** [typescript-eslint/typescript-eslint#1845]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": [2, {"allowExpressions": true}],
    "@typescript-eslint/explicit-module-boundary-types": [2]
  }
}
```

```TS
export function foo(): (n: number) => string {
  return (n) => String(n)
}
```

**Expected Result**

Since the module boundary (`foo`) is fully typed, I would not expect to have to type the inner function expression.

**Actual Result**

`explicit-function-return-type` permits this code when the `allowExpressions` option is enabled. But `explicit-module-boundary-types` complains that the argument and return value of the inner function expression are untyped, and the rule doesn't have a corresponding expressions option.

**Additional Info**

n/a

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.26.0` |
| `@typescript-eslint/parser`        | `2.26.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.11.0` |
| `npm`                              | `6.14.4` |

**Comment by angelayi** [comment]
Hi! I'm pretty new to contributing to open source projects, but I was wondering if I could try working on this issue?

**Comment by bradzacher** [comment]
Feel free to dig in!
https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#pull-requests

**Comment by ifiokjr** [comment]
I just came here to post the same issue after upgrading to `3.0.0`.

Is anyone currently working on this?

**Comment by bradzacher** [comment]
feel free!

currently the rule recursively checks functions in case you are doing a higher-order function: `const x = () => () => (): string => 'foo'`. 

Should be a simple enough fix though - it shouldn't do this recursive check if the parent function has a type annotation.

**Comment by ifiokjr** [comment]
Okay, I've pushed up a potential fix. 

It's a really high quality codebase. I've never worked with  AST's before but the logic and tests here are structured really well. I think TypeScript together with https://astexplorer.com really helps.

**Comment by essenmitsosse** [comment]
Just installed 3.1.0. The example function from the opening post still gives the same error for me.

```
export function foo(): (n: number) => string {
  return (n) => String(n)
}
```
"Missing return type on function."
"Argument 'n' should be typed."

Already reinstalled all packages and cleared cache.

package | version
-- | --
@typescript-eslint/eslint-plugin | 3.1.0
@typescript-eslint/parser | 3.1.0
TypeScript | 3.9.3
ESLint | 7.1.0
node | 13.14.0
npm | 6.14.4

**Comment by bradzacher** [comment]
There is a passing test for this exact case in master, so it should be working fine.

https://github.com/typescript-eslint/typescript-eslint/blob/c86e2a235372149db9b1700d39c2145e0ce5221a/packages/eslint-plugin/tests/rules/explicit-module-boundary-types.test.ts#L471-L477

Could you please:
- file a new issue, as per our [contributing guideline](https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#commenting)
- create a repro repo to help us investigate.
